src/ucp/core/ucp_worker.c

```diff
@@ -2453,45 +2455,87 @@ static int ucp_worker_keepalive_is_enabled(ucp_worker_h worker)
+    if (worker->keepalive.iter == &worker->all_eps) {
+        /* if next list item points to all_eps then step one more time */
+        worker->keepalive.iter = worker->keepalive.iter->next;
+    }
+
+    ucs_assert(worker->keepalive.iter != &worker->all_eps);
+    ep                         = ucp_worker_keepalive_current_ep(worker);
+    worker->keepalive.lane_map = ((ep->cfg_index != UCP_WORKER_CFG_INDEX_NULL) &&
+                                  !(ep->flags & UCP_EP_FLAG_FAILED)) ?
+                                 ucp_ep_config(ep)->key.ep_check_map : 0;
+}
+
 static unsigned ucp_worker_keepalive_progress(void *arg)
 {
     ucp_worker_h worker = (ucp_worker_h)arg;
     ucs_time_t now      = ucs_get_time();
-    unsigned ep_keepalive_count;
-    ucp_ep_ext_gen_t *ep_ext;
     ucs_list_link_t *iter_begin;
+    ucp_ep_h ep;
 
     if (ucs_likely((now - worker->keepalive.last_round) <
                    worker->context->config.ext.keepalive_timeout)) {
         return 0;
     }
 
-    worker->keepalive.last_round = now;
-    ep_keepalive_count           = 0;
-    iter_begin                   = worker->keepalive.iter;
+    if (ucs_unlikely(ucs_list_is_empty(&worker->all_eps))) {
+        ucs_assert(worker->keepalive.iter == &worker->all_eps);
+        worker->keepalive.ep_count = 0;
+        goto out;
+    }
+
+    if (ucs_unlikely(worker->keepalive.iter == &worker->all_eps)) {
+        ucp_worker_keepalive_next_ep(worker);
+    }
+
+    iter_begin = worker->keepalive.iter;
     /* use own loop for elements because standard for_each skips
      * head element */
     /* TODO: use more optimal algo to enumerate EPs to keepalive
      * (linked list) */
     do {
-        ep_ext = ucs_list_next(worker->keepalive.iter,
-                               ucp_ep_ext_gen_t, ep_list);
-        worker->keepalive.iter = &ep_ext->ep_list;
-        if (worker->keepalive.iter == &worker->all_eps) {
-            continue;
+        ep = ucp_worker_keepalive_current_ep(worker);
+        ucp_ep_do_keepalive(ep, &worker->keepalive.lane_map);
+        if (worker->keepalive.lane_map != 0) {
+            /* in case if EP has no resources to send keepalive message
+             * then just return without update of last_round timestamp,
+             * on next progress iteration we will continue from this point */
+            goto out_no_resources;
         }
 
-        ep_keepalive_count += ucp_ep_do_keepalive(ucp_ep_from_ext_gen(ep_ext));
-    } while ((iter_begin != &ep_ext->ep_list) &&
-             (ep_keepalive_count < worker->context->config.ext.keepalive_num_eps));
+        worker->keepalive.ep_count++;
+        ucp_worker_keepalive_next_ep(worker);
+    } while ((iter_begin != worker->keepalive.iter) &&
+             (worker->keepalive.ep_count < worker->context->config.ext.keepalive_num_eps));
 
-    if (ep_keepalive_count == 0) {
+    worker->keepalive.last_round = now;
+
+out:
+    if (worker->keepalive.ep_count == 0) {
```

Comment:
seems "uct_worker_progress_unregister_safe" is done only if list is empty
so maybe move it instead of line 2498 (inside the "if"), then remove label "out" and also remove line 2535
then, can remove the label "out_no_resources" and replace "goto out_no_resources" by "break"

Reply:
```break``` will not work because we have to reset timeout/ep_count only in case if no ```IN_PROGRESS``` status returned

Reply:
right